### PR TITLE
[Feature] support mix precision training for resnet

### PR DESCRIPTION
**Pull request by typhoonzero**
For reference: https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html

- update conv cudnn bwd for fp16
- more accurate mode for fp16 softmax (need future log_softmax support)
- The optimizer change is for loss and gradient scaling in: https://github.com/PaddlePaddle/models/pull/1533

TODO: support log_softmax and nll_loss (need test full training)

Previous work: https://github.com/PaddlePaddle/Paddle/pull/14992

### CODE REVIEW

File: paddle/fluid/framework/details/build_strategy.cc [hunk 1]
```diff
@@ -67,7 +67,7 @@ class ParallelExecutorPassBuilder : public ir::PassBuilder {
     context->endpoints_ = strategy_.trainers_endpoints_;
     context->trainer_id_ = strategy_.trainer_id_;
     PADDLE_ENFORCE(strategy_.trainer_id_ >= 0, "trainer_id_ >= 0");
-    if (strategy_.trainer_id_ > 0) {
+    if (strategy_.trainer_id_ > 0 && strategy_.trainers_endpoints_.size() > 0) {
```
Comment (Yancey0623) [inline paddle/fluid/framework/details/build_strategy.cc]: Duplicated condition with L71 ?

File: paddle/fluid/operators/conv_cudnn_op.cu.cc [hunk 2]
```diff
@@ -297,6 +297,21 @@ class CUDNNConvGradOpKernel : public framework::OpKernel<T> {
...
+    if (dev_ctx.GetComputeCapability() >= 70 &&
+        std::type_index(typeid(T)) ==
+            std::type_index(typeid(platform::float16))) {
+      CUDNN_ENFORCE(platform::dynload::cudnnSetConvolutionMathType(
+          cudnn_conv_desc, CUDNN_TENSOR_OP_MATH));
+      VLOG(5) << "use cudnn_tensor_op_math for bwd";
+    } else {
+      CUDNN_ENFORCE(platform::dynload::cudnnSetConvolutionMathType(
+          cudnn_conv_desc, CUDNN_DEFAULT_MATH));
+      VLOG(5) << "NOT use cudnn_tensor_op_math for bwd";
```
Comment (Yancey0623) [inline paddle/fluid/operators/conv_cudnn_op.cu.cc]: bwd => backward ?

File: paddle/fluid/operators/math/reductions.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chengduoZH) [inline paddle/fluid/operators/math/reductions.h]: 2016->2018

File: paddle/fluid/operators/math/reductions.h [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    int gid = threadIdx.x + blockIdx.x * blockDim.x;
+    if (gid < num_rows) out[gid] = in[gid];
+    return;
+  }
+
+  value_type sum = initVal;
+  int col = lane;
+
+  if (row < num_rows && col < num_cols) {
+    sum = in[row * num_cols + col];
```
Comment (chengduoZH) [inline paddle/fluid/operators/math/reductions.h]: Why do you use `assign` here? if so, the `initVal ` is meaningless.

File: paddle/fluid/operators/math/reductions.h [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    col += 32;
+    for (; col < num_cols; col += 32) {
+      sum = op(sum, in[row * num_cols + col]);
+    }
+  }
+
+  typedef cub::WarpReduce<value_type> WarpReduce;
+
+  __shared__ typename WarpReduce::TempStorage temp_storage;
```
Comment (chengduoZH) [inline paddle/fluid/operators/math/reductions.h]: Why do you use `WarpReduce ` but not `BlockReduce ` here?

File: paddle/fluid/operators/math/reductions.h [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+template <typename T, typename OUT_T, typename IN_T, typename Op>
+void LaunchRowReduction(const platform::CUDADeviceContext* ctx, OUT_T out,
+                        IN_T in, int num_rows, int num_cols, Op op, T init) {
+  if (num_cols < 1024) {
+    const int threads_per_block = 128;
+    const int warps_per_block = threads_per_block / 32;
+    int num_blocks = (num_rows + warps_per_block - 1) / warps_per_block;
+
+    RowReduceKernel<<<num_blocks, threads_per_block, 0, ctx->stream()>>>(
+        in, out, num_rows, num_cols, op, init);
```
Comment (chengduoZH) [inline paddle/fluid/operators/math/reductions.h]: This computation of `num_blocks ` and `threads_per_block ` is too complicated, and it is confusing.

File: paddle/fluid/operators/math/reductions.h [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    auto success = cub::DeviceSegmentedReduce::Reduce(
+        i == 0 ? nullptr : temp_storage.mutable_data<uint8_t>(ctx->GetPlace()),
+        temp_storage_bytes, in, out, num_rows, transform_iter,
+        transform_iter + 1, op, init, ctx->stream());
+
+    PADDLE_ENFORCE_EQ(success, 0, "CUB segmented reduce error");
+    if (i == 0) {
+      temp_storage.Resize({static_cast<int64_t>(temp_storage_bytes)});
+      // do allocate
+      temp_storage.mutable_data<uint8_t>(ctx->GetPlace());
```
Comment (chengduoZH) [inline paddle/fluid/operators/math/reductions.h]: Note that the `temp_storage` is a stack memory, and it will be freed once the current function return, but the CUDA calling is async.
You can use Temporary Allocator here.

File: python/paddle/fluid/optimizer.py [hunk 8]
```diff
@@ -291,18 +292,41 @@ def _process_distribute_lookuptable(self, param_grads, loss,
...
+                 no_grad_set=None,
+                 user_params_grads=None):
         """Add operations to minimize `loss` by updating `parameter_list`.
 
         This method combines interface `append_backward()` and
         `create_optimization_pass()` into one.
         """
-        params_grads = append_backward(loss, parameter_list, no_grad_set,
-                                       [error_clip_callback])
+        if not self._backward_called:
```
Comment (Yancey0623) [inline python/paddle/fluid/optimizer.py]: Is `_backward` interface is public?

File: paddle/fluid/operators/math/softmax.cu [hunk 9]
```diff
@@ -108,6 +112,119 @@ template class SoftmaxGradFunctor<platform::CUDADeviceContext, double>;
...
+  max_logits.Resize(max_dims);
+  sum_probs.Resize(X->dims());
+  auto* max_logits_data = max_logits.mutable_data<T>(ctx.GetPlace());
+  auto* logits_data = X->data<T>();
+  auto* sum_probs_data = sum_probs.mutable_data<ACCURATE_T>(ctx.GetPlace());
+  int num_rows = X->dims()[0];
+  int num_cols = X->dims()[1];
+
+  // Use ACCURATE_T for internal exp space
+  Tensor acc_out_tensor;
```
Comment (chengduoZH) [inline paddle/fluid/operators/math/softmax.cu]: `acc_out_tensor ` is a stack memory.

**Comment by panyx0718** [comment]
Is there a mix precision training design doc?

File: paddle/fluid/operators/math/reductions.h (hunk 5 above)
Comment (typhoonzero) [reply paddle/fluid/operators/math/reductions.h]: Note that `RowReduceKernel` will be used only when `num_cols < 1024` in this case, reduce will be called under maximum 32 lanes and it will be faster by using `WarpReduce` through those threads. If `num_cols >= 1024`, will use `DeviceSegmentedReduce`

File: paddle/fluid/operators/math/reductions.h (hunk 6 above)
Comment (typhoonzero) [reply paddle/fluid/operators/math/reductions.h]: `(a + b - 1) / b` is same to divup( a / b)

File: paddle/fluid/framework/details/build_strategy.cc (hunk 1 above)
Comment (typhoonzero) [reply paddle/fluid/framework/details/build_strategy.cc]: enforce is to ensure `trainer_id_ < endpoints_` and only when we provided `trainers_endpoints_` need to check, which is used for gpu sparse collective server only.

File: python/paddle/fluid/optimizer.py (hunk 8 above)
Comment (typhoonzero) [reply python/paddle/fluid/optimizer.py]: not recommended for user to call this. only used for mix precision for optimize using fp32.

**Comment by typhoonzero** [comment]
@panyx0718 The implementation is exactly the same as https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html

File: python/paddle/fluid/optimizer.py [hunk 10]
```diff
@@ -291,18 +292,41 @@ def _process_distribute_lookuptable(self, param_grads, loss,
...
+            self._backward(loss, startup_program, parameter_list, no_grad_set)
+
+        # optimize user specified parameters
+        if user_params_grads:
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: where is user_param_grads from? If it is available, why call _backward? Will backward be created twice?

**Comment by panyx0718** [comment]
How do you define a network with partly float16 using layers API?

File: python/paddle/fluid/optimizer.py (hunk 10 above)
Comment (typhoonzero) [reply python/paddle/fluid/optimizer.py]: Usage is: https://github.com/PaddlePaddle/models/pull/1533/files#diff-75955d9a0dea899fef1dc507808feb45R255

File: paddle/fluid/operators/conv_cudnn_op.cu.cc [hunk 11]
```diff
@@ -297,6 +297,21 @@ class CUDNNConvGradOpKernel : public framework::OpKernel<T> {
     cudnnFilterDescriptor_t cudnn_filter_desc = filter_desc.descriptor<T>(
         layout, framework::vectorize2int(filter->dims()), groups);
 
+#if CUDA_VERSION >= 9000 && CUDNN_VERSION_MIN(7, 0, 1)
+    // Enable Tensor Core for cudnn backward
+    if (dev_ctx.GetComputeCapability() >= 70 &&
+        std::type_index(typeid(T)) ==
+            std::type_index(typeid(platform::float16))) {
+      CUDNN_ENFORCE(platform::dynload::cudnnSetConvolutionMathType(
```
Comment (panyx0718) [inline paddle/fluid/operators/conv_cudnn_op.cu.cc:305]: how much faster is this?

File: paddle/fluid/operators/math/softmax.cu [hunk 12]
```diff
@@ -108,6 +113,120 @@ template class SoftmaxGradFunctor<platform::CUDADeviceContext, double>;
 template class SoftmaxGradFunctor<platform::CUDADeviceContext,
                                   platform::float16>;
 
+// NOTE: upgrade value type for float16 and float32
+static __device__ __forceinline__ float real_exp(platform::float16 x) {
```
Comment (panyx0718) [inline paddle/fluid/operators/math/softmax.cu]: Should be called "cast_to_float32_exp". Not sure what does "real" mean

File: paddle/fluid/operators/math/softmax.cu [hunk 13]
```diff
@@ -108,6 +113,120 @@ template class SoftmaxGradFunctor<platform::CUDADeviceContext, double>;
...
+  return ::Eigen::numext::exp(static_cast<float>(x));
+}
+static __device__ __forceinline__ double real_exp(float x) {
```
Comment (panyx0718) [inline paddle/fluid/operators/math/softmax.cu]: same as above

File: paddle/fluid/operators/softmax_op.h [hunk 14]
```diff
@@ -45,6 +46,30 @@ class SoftmaxKernel : public framework::OpKernel<T> {
   }
 };
 
+// specialized for fp16
+#ifdef PADDLE_WITH_CUDA
+template <>
+class SoftmaxKernel<platform::CUDADeviceContext, platform::float16>
```
Comment (panyx0718) [inline paddle/fluid/operators/softmax_op.h]: why not just fluid.cast to float32 and run float32 softmax?

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 15]
```diff
@@ -395,6 +409,11 @@ class SoftmaxWithCrossEntropyCUDAKernel : public framework::OpKernel<T> {
 
     auto soft_label = context.Attr<bool>("soft_label");
     auto ignore_index = context.Attr<int>("ignore_index");
+    bool half_to_float = false;
+    if (std::type_index(typeid(T)) == std::type_index(typeid(fp16))) {
+      half_to_float = true;
```
Comment (panyx0718) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: looks hacky....

File: python/paddle/fluid/optimizer.py [hunk 16]
```diff
@@ -292,11 +293,28 @@ def _process_distribute_lookuptable(self, param_grads, loss,
...
+                                            callbacks)
+        return self.params_grads
+
     def minimize(self,
                  loss,
                  startup_program=None,
                  parameter_list=None,
-                 no_grad_set=None):
+                 no_grad_set=None,
+                 user_params_grads=None):
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: no comments? Doesn't fit well with `parameter_list`

File: python/paddle/fluid/optimizer.py [hunk 17]
```diff
@@ -321,8 +339,15 @@ def minimize(self,
             optimize_ops = self._create_optimization_pass(params_grads, loss,
                                                           startup_program)
         else:
-            params_grads = append_backward(loss, parameter_list, no_grad_set,
-                                           [error_clip_callback])
+            if not self._backward_called:
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: what's the point here? Why not just call minimize once?

File: python/paddle/fluid/optimizer.py [hunk 18]
```diff
@@ -321,8 +339,15 @@ def minimize(self,
...
+                self._backward(loss, startup_program, parameter_list,
+                               no_grad_set)
+
+            # optimize user specified parameters
+            if user_params_grads:
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: take a look at tf.Optimizer.apply_gradients, you might have better idea

File: paddle/fluid/operators/math/softmax.cu (hunk 12 above)
Comment (dzhwinter) [reply paddle/fluid/operators/math/softmax.cu]: this should be put into float16.h and extend the std namespace.

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu [hunk 19]
```diff
@@ -58,8 +63,17 @@ __global__ void SoftCrossEntropyGradientKernel(T* logit_grad,
 
 }  // namespace
 
+static __device__ __forceinline__ platform::float16 real_exp(
```
Comment (dzhwinter) [inline paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: same here.

**Comment by dzhwinter** [comment]
Basically, I thought that we can have a optimizer wrapper, then different optimizer can be unified into the same wrapper. Maybe we can modify the optimizer interface.

File: python/paddle/fluid/optimizer.py (hunk 18 above)
Comment (typhoonzero) [reply python/paddle/fluid/optimizer.py]: Well the actual implementation is the same, though might need to seperate APIs in another PR would be more cleaner.

File: paddle/fluid/operators/math/softmax.cu (hunk 12 above)
Comment (typhoonzero) [reply paddle/fluid/operators/math/softmax.cu]: @dzhwinter fp32 still need a way to return double result here, for current usage I'll keep it here for now and refine this later.

**Comment by typhoonzero** [comment]
> Basically, I thought that we can have a optimizer wrapper, then different optimizer can be unified into the same wrapper. Maybe we can modify the optimizer interface.

Thanks, I'll try to do that in next PR to refine the optimzier API.

File: paddle/fluid/operators/softmax_with_cross_entropy_op.cu (hunk 19 above)
Comment (typhoonzero) [reply paddle/fluid/operators/softmax_with_cross_entropy_op.cu]: Just renamed `exp` and removed `real_log`

File: python/paddle/fluid/optimizer.py [hunk 20]
```diff
@@ -275,38 +269,121 @@ def _process_distribute_lookuptable(self, param_grads, loss,
...
+                    type='sgd',
+                    inputs={
+                        "Param": table_param,
+                        "Grad": table_grad,
+                        "LearningRate": self._create_param_lr(param_and_grad)
+                    },
+                    outputs={"ParamOut": param_and_grad[0]})
         return new_param_grads, (table_param, table_grad), sgd_op
 
+    def backward(self,
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py:286]: currently, this API seems the same as `append_backward`. It seems not no needed now. Let's add it when necessary.

File: paddle/fluid/operators/math/softmax.h [hunk 21]
```diff
@@ -49,6 +49,14 @@ class SoftmaxGradCUDNNFunctor {
                   const framework::Tensor* Y, const framework::Tensor* y_grad,
                   framework::Tensor* x_grad);
 };
+
+template <typename T, typename ACCURATE_T>
+class SoftmaxCudaAccurateFunctor {
```
Comment (panyx0718) [inline paddle/fluid/operators/math/softmax.h]: Is this still used?

File: python/paddle/fluid/optimizer.py [hunk 22]
```diff
@@ -70,6 +70,7 @@ def __init__(self, learning_rate, regularization=None, name=None):
         # {accum_name : { paramter_name : accumulator_for_parameter, ...}, ...}
         self._accumulators = defaultdict(lambda: dict())
         self.helper = None
+        self._backward_called = False
```
Comment (panyx0718) [inline python/paddle/fluid/optimizer.py]: still needed?

File: python/paddle/fluid/optimizer.py (hunk 20 above)
Comment (typhoonzero) [reply python/paddle/fluid/optimizer.py:286]: Need this so that user do not need to import `append_backward` somewhere else and add `error_clip_callback` manually.

File: python/paddle/fluid/optimizer.py (hunk 22 above)
Comment (typhoonzero) [reply python/paddle/fluid/optimizer.py]: Thanks, done.

File: paddle/fluid/operators/math/softmax.h (hunk 21 above)
Comment (typhoonzero) [reply paddle/fluid/operators/math/softmax.h]: Thanks, done.

**Review by panyx0718 — APPROVED** [review]
